Detect expired messages in Kafka. Log and set a guage.

vrajat opened the pull request:
Pinot may take multiple hours between polling a partition in a Kafka topic. One specific example is that Pinot took a long time to flush a segment to disk. In the meantime, messages in Kafka can expire if message retention time is small. 
If `auto.offset.reset` is set to `smallest`, then Kafka will silently move the offset to the first available message leading to data loss. 
`RealtimeSegmentValidationManager` is a cron that runs every hour and detects where the offset of a segment in zookeeper is in the past when compared to the smallest offset in Kafka. However since it runs every hour, it may miss the data loss if it happens between runs.

This commit has the following contributions:
1. Check the startOffset to beginOffset. If `startOffset` < `beginOffset`, then log the condition as well as set a meter to 1.
2. Augment debug API to also report data loss errors. 
3. Write a test that simulates the data loss and checks debug APIs to ensure that the simulation succeeded.

ege-st commented on pinot-core/src/main/java/org/apache/pinot/core/data/manager/realtime/RealtimeSegmentDataManager.java:
```diff
@@ -1648,6 +1648,15 @@ private void makeStreamConsumer(String reason) {
         _streamConsumerFactory.createPartitionGroupConsumer(_clientId, _partitionGroupConsumptionStatus);
+    try {
+      _partitionGroupConsumer.validateStreamState(_currentOffset);
+      _serverMetrics.setValueOfPartitionGauge(_tableNameWithType, 0,
```
So it's good that we have a gauge per table (because otherwise, a table which is healthy could reset the gauge from `1` to `0` if it executes this check _after_ an unhealthy table set the gauge to `1`).

However, what happens if there are multiple partitions on a table and one partition's segment sets this gauge to `1`, then a different partition's segment executes this code and resets the gauge from `1` to `0`? Put another way: if a table has 3 partitions and 1 of them fails `validateStreamState` but the other two pass `validateStreamState` then it's very likely that the healthy partitions will mask the unhealthy partition.

We need to make sure that if there is a segment that fails this validation this gauge is set to `1` until that segment is fixed or there is some other manual intervention.

vrajat replied:
The gauge for a specific partition. `setValueOfParitionGauge` takes a partition id as one of the parameters.

vrajat replied:
Going back on this as well. Metrics/alerts per partition is not allowed anymore. So the metric/alert is for a table. Logs have more information.

ege-st commented on pinot-core/src/main/java/org/apache/pinot/core/data/manager/realtime/RealtimeSegmentDataManager.java:
```diff
@@ -1648,6 +1648,15 @@ private void makeStreamConsumer(String reason) {
+      _partitionGroupConsumer.validateStreamState(_currentOffset);
+      _serverMetrics.setValueOfPartitionGauge(_tableNameWithType, 0,
+              ServerGauge.INVALID_REALTIME_STREAM_STATE_EXCEPTION, 0);
+    } catch (PermanentConsumerException pce) {
```
Should we catch this exception here or bubble it up to the next level?

vrajat replied:
I wasnt sure if the ingest should be stopped. After a couple of conversations the feedback is that it should be stopped. I'll bubble up the exception.

vrajat replied:
Going back on this. On more discussion, it is a bad idea to stop ingestion as that will cause more data loss. So the decision right now is to alert but not stop. Also not that the same condition is checked in `RealtimeDataValidationManager` and that increases a meter but does not attempt to stop. So there is a precedent.

ege-st commented on pinot-plugins/pinot-stream-ingestion/pinot-kafka-2.0/src/main/java/org/apache/pinot/plugin/stream/kafka20/KafkaPartitionLevelConsumer.java:
```diff
@@ -46,6 +50,20 @@ public KafkaPartitionLevelConsumer(String clientId, StreamConfig streamConfig, i
+            _consumer.beginningOffsets(Collections.singletonList(_topicPartition));
+
+    final long beginningOffset = beginningOffsets.getOrDefault(_topicPartition, 0L);
+    if (startOffset < beginningOffset) {
```
Is `startOffset > beginningOffset` an acceptable scenario?

vrajat replied:
Yes. That is the common scenario. [KafkaConsumer::position](https://kafka.apache.org/0110/javadoc/org/apache/kafka/clients/consumer/KafkaConsumer.html#position(org.apache.kafka.common.TopicPartition)) specifies the next record that is somewhere in the middle or end of the stream. Also `startOffset` is usually the same as `KafkaConsumer::position` if the consumer is keeping up.

ege-st replied:
Coming back to this discussion, I realized my confusion: in this case Beginning Offset is the earliest offset in the Kafka topic and Start Offset is where the new Consuming Segment will start reading from.  

I had been thinking in terms of the last offset in the immediately preceding segment.

Jackie-Jiang reviewed (commented):
We should be able to emit metric within `PartitionGroupConsumer.fetchMessages()` when the start offset is not available (e.g. the asked offset range is not fully returned).

vrajat commented:
> We should be able to emit metric within `PartitionGroupConsumer.fetchMessages()` when the start offset is not available (e.g. the asked offset range is not fully returned).

@Jackie-Jiang Are you suggesting to run this check in in `RealtimeSegmentDataManager::consumeLoop` effectively instead of when `PartitionGroupConsumer::start` is called? Will that cause too many checks? 

One doubt I have is whether the consumer may fall behind even after the consumer was started. In that case it is better to move the check to `consumeLoop` or `fetchMessages`.

swaminathanmanish commented:
> > We should be able to emit metric within `PartitionGroupConsumer.fetchMessages()` when the start offset is not available (e.g. the asked offset range is not fully returned).
> 
> @Jackie-Jiang Are you suggesting to run this check in in `RealtimeSegmentDataManager::consumeLoop` effectively instead of when `PartitionGroupConsumer::start` is called? Will that cause too many checks?
> 
> One doubt I have is whether the consumer may fall behind even after the consumer was started. In that case it is better to move the check to `consumeLoop` or `fetchMessages`.

Yes I think its possible that records in kafka can expire while consumeLoop is consuming from stream (beyond start phase). Filling a realtime segment can take several hours during which records may expire. 

@Jackie-Jiang - Is your intention to add the check to `PartitionGroupConsumer.fetchMessages()` for completeness of the check as well as reuse across stream fetchers ?

ege-st commented:
It is possible for messages to get expired in the middle of the lifetime of a consuming segment. One way this could happen is if the ingestion on the table is paused for a period longer than the Source's retention, when ingestion is resumed it would lead to a gap in messages within the segment.

ege-st commented:
> > We should be able to emit metric within `PartitionGroupConsumer.fetchMessages()` when the start offset is not available (e.g. the asked offset range is not fully returned).
> 
> @Jackie-Jiang Are you suggesting to run this check in in `RealtimeSegmentDataManager::consumeLoop` effectively instead of when `PartitionGroupConsumer::start` is called? Will that cause too many checks?
> 
> One doubt I have is whether the consumer may fall behind even after the consumer was started. In that case it is better to move the check to `consumeLoop` or `fetchMessages`.

I don't think there will be a noticeable performance impact if the check is done in `fetchMessages` or `consumeLoop` and executed for each message batch. The performance cost should be relatively small compared to everything else that we do in the `consumeLoop`. But it would be awesome to do a performance test just to make sure.

Jackie-Jiang commented:
The overhead should be minimal if we don't need to make extra remote calls. I'd assume `PartitionGroupConsumer.fetchMessages()` already have the information we needed to perform the check

vrajat commented:
A simple script to show message expiration in kafka

Assume all the components are installed
- Kafka
- Kafka CLIs
- Faker

```
# Create a topic with retention of 1m
kafka_2.13-3.6.1/bin/kafka-topics.sh --bootstrap-server localhost:9092 --create --config retention.ms=60000 --topic cli_profiles

# Produce 5 rows
faker -r 5 profile | nl -w2 -v 10 -s':' | ./kafka_2.13-3.6.1/bin/kafka-console-producer.sh --bootstrap-server localhost:9092 --property parse.key=true --property key.separator=":" --topic cli_profile

# Example row:
1:{'job': 'Health service manager', 'company': 'Wall, Ballard and Anderson', 'ssn': '729-80-0324', 'residence': '42764 Porter Mountains\nStephenborough, NV 63641', 'current_location': (Decimal('56.096589'), Decimal('-156.143961')), 'blood_group': 'AB+', 'website': ['http://serrano.org/', 'https://huerta-kidd.com/'], 'username': 'martinezchristopher', 'name': 'Linda Taylor', 'sex': 'F', 'address': '0509 Angela Road\nEast Sandraport, MP 39203', 'mail': 'regina54@gmail.com', 'birthdate': datetime.date(1923, 9, 8)}

# Consume rows after a sleep of 90 seconds. None of these rows should be available.
# kafka-console-consumer.sh is in wait state without printing any rows.
sleep 90
./kafka_2.13-3.6.1/bin/kafka-console-consumer.sh --bootstrap-server localhost:9092 --property print.key=true --property key.separator=":" --topic cli_profile
```

vrajat commented:
I've added `RealtimeSlowConsumer.java` as a temporary reproducer while I work on an integration test.

ege-st commented:
So to make sure I understand: we'll essentially have a gauge per partition per table. And if a new consuming segment is started where the start offset < earliest offset still in Kafka the Gauge will be set to `1` for that partition of that table.  

That gauge will remain at `1` until the next time a consuming segment is created and it will be set to `0` if start offset >= the earliest offset in Kafka.

vrajat commented:
> So to make sure I understand: we'll essentially have a gauge per partition per table. And if a new consuming segment is started where the start offset < earliest offset still in Kafka the Gauge will be set to `1` for that partition of that table.
> 
> That gauge will remain at `1` until the next time a consuming segment is created and it will be set to `0` if start offset >= the earliest offset in Kafka.

good questions here and we should discuss what is the right thing to do. 
* I have changed the code to a metric per table. As part of alert clean up, SRE team has requested to remove per partition alerts. So the on-call has to refer to logs to get the lagging partitions. We are also changing existing alerts to be table specific.
* Another important question is if ingestion should stop. One worry is that stopping ingestion may make data loss worse as more messages may expire.
* If ingestion continues, then when and how should the metric be reset?

vrajat commented:
@Jackie-Jiang @ege-st Can you please re-review this PR? The major changes since the last changes are:
1. Cleaned up the check in `RealtimeSegmentDataManager` to reuse member objects to check offsets.
2. Report validation errors from both `RealtimeSegmentDataManager` and `RealtimeSegmentValidationManager` to Debug APIs.
3. Write a test that simulates this scenario. It also waits for Kafka to fast forward and the error messages to appear in the debug info.

Jackie-Jiang commented on pinot-tools/src/main/java/org/apache/pinot/tools/RealTimeSlowConsumer.java:
```diff
@@ -0,0 +1,142 @@
+
+    @Override
+    protected Map<String, String> getDefaultStreamTableDirectories() {
+        return ImmutableMap.<String, String>builder()
```
The indentation is incorrect. Please apply [Pinot Style](https://docs.pinot.apache.org/developers/developers-and-contributors/code-setup#set-up-ide) and reformat the changes

vrajat replied:
More existential question: Should I keep this quickstart or discard it? It is helpful to play around for this specific scenario but I dont see a use for it beyond that. I do have an integration test.

Jackie-Jiang commented on pinot-core/src/main/java/org/apache/pinot/core/data/manager/realtime/RealtimeSegmentDataManager.java:
```diff
@@ -889,6 +892,39 @@ public Map<String, PartitionLagState> getPartitionToLagState(
+   * @param startOffset The offset of the first message desired, inclusive
+   */
+  private void validateStartOffset(StreamPartitionMsgOffset startOffset) {
+    if (_partitionMetadataProvider == null) {
+      createPartitionMetadataProvider("validateStartOffset");
+    }
+
+    try {
+      StreamPartitionMsgOffset streamSmallestOffset = _partitionMetadataProvider.fetchStreamPartitionOffset(
+          OffsetCriteria.SMALLEST_OFFSET_CRITERIA,
+          /*maxWaitTimeMs=*/5000
+      );
```
Use `fetchEarliestStreamOffset(5000)`?

Jackie-Jiang commented on pinot-core/src/main/java/org/apache/pinot/core/data/manager/realtime/RealtimeSegmentDataManager.java:
```diff
@@ -889,6 +892,39 @@ public Map<String, PartitionLagState> getPartitionToLagState(
+            + ") is older than topic's beginning offset(" + streamSmallestOffset + ")";
+        _segmentLogger.error(message);
+        _realtimeTableDataManager.addSegmentError(_segmentNameStr,
+            new SegmentErrorInfo(String.valueOf(now()), message, "")
```
```suggestion
            new SegmentErrorInfo(now()), message, null)
```

vrajat replied:
Made this change locally.

Jackie-Jiang commented on pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/realtime/PinotLLCRealtimeSegmentManager.java:
```diff
@@ -1711,4 +1725,14 @@ String moveSegmentFile(String rawTableName, String segmentName, String segmentLo
     return URIUtils.getUri(_controllerConf.getDataDir(), rawTableName, URIUtils.encode(segmentName));
   }
+
+  public Map<String, SegmentErrorInfo> getSegmentErrors(String tableNameWithType) {
```
Why not storing a 2 level map: `Map<String, Map<String, SegmentErrorInfo>>`

vrajat replied:
Hmm - I was following the example of `protected LoadingCache<Pair<String, String>, SegmentErrorInfo> _errorCache;` but the same constraint is not here. I'll change it.

vrajat replied:
One issue with a 2-level map is that it is harder to control the size. Getting the total size of the errors will require iteration.

Jackie-Jiang commented on pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/realtime/PinotLLCRealtimeSegmentManager.java:
```diff
@@ -1339,7 +1346,14 @@ private StreamPartitionMsgOffset selectStartOffset(OffsetCriteria offsetCriteria
+                        + " of table: " + tableName;
+
+        _errorCache.put(Pair.of(tableName, segmentName),
+            new SegmentErrorInfo(String.valueOf(System.currentTimeMillis()), message, ""));
```
```suggestion
            new SegmentErrorInfo(System.currentTimeMillis()), message, null));
```

vrajat replied:
Made this change locally.

Jackie-Jiang commented on pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/realtime/PinotLLCRealtimeSegmentManager.java:
```diff
@@ -140,6 +142,8 @@ public class PinotLLCRealtimeSegmentManager {
   // Max time to wait for all LLC segments to complete committing their metadata while stopping the controller.
   private static final long MAX_LLC_SEGMENT_METADATA_COMMIT_TIME_MILLIS = 30_000L;
 
+  private Map<Pair<String, String>, SegmentErrorInfo> _errorCache;
```
This is not really a cache because we will never clean up the entries as of now. Do you plan to add the cleanup logic in this PR? If not, let's add a TODO at least

swaminathanmanish replied:
Im not sure if this is meant to be cleaned up, since we need keep it around for the debug api calls ?

vrajat replied:
I should limit its size. I am thinking of using a `LinkedHashMap` to limit its entry. Similar to https://docs.oracle.com/javase/8/docs/api/java/util/LinkedHashMap.html#removeEldestEntry-java.util.Map.Entry-

Is that OK? 

However, these methods are not synchronized. I can use [Collections.synchronizedMap](https://docs.oracle.com/javase/8/docs/api/java/util/Collections.html#synchronizedMap-java.util.Map-) to wrap the LinkedHashMap. 

Also another minor improvement is to change the signature to `getSegmentErrors(String tableName, String segmentName)`. Then iterators and streams are not required and all the access will be through `get` which is synchronized.

ege-st replied:
Just double checking my understanding of the error cache.  It's a map from each (table, segment) pair that is on this server to the most recent error message that was seen for that table/segment? In other words, for each server, we'll see the most recent error on each segment on that server. 

1. Longer term is how to manage noisy errors vs not-noisy errors.  For example: if there's an error with missing offsets (which you're monitoring for in this PR) and a decoding error on 1/5 messages, the decoding error will flood the cache and block out the Offset Error from being seen.
2. What happens when a table/segment is deleted or moved? The error cache will still have the non-existent segments and provide invalid information.  We have this issue with Ingestion Lag metrics and it's frequently causing false alerts and issues. If this happens multiple times then we can wind up with many servers reporting errors for the same segment which will be confusing during investigations.
3. If you limit the size of this map, then it still needs to support all the extant segments that are on a server: so I'm not sure setting a fixed limit will work b/c how many segments a single server can have is not, so far as I know, strictly limited. So how can we determine what the max size should be?

vrajat replied:
One point is that this is for debug APIs only. Errors from this list is not exposed to the user. The official user interface is metrics and logs. Based on the fact that the error list is for debugging only:


1.  High error rate: Limiting the size is a good idea so that it doesnt use up memory. Losing errors is OK as this is not the source of truth and only for debugging. Right now there is only one error source. If there are too many, the important aspect is that a data loss occurred.
2. Segment lifecycle: Since it is not the source of truth and used by devs only, mismatch is OK. The main contribution of the PR is the metric to track data loss and the log with all the necessary info.

Jackie-Jiang commented on pinot-core/src/main/java/org/apache/pinot/core/data/manager/realtime/RealtimeSegmentDataManager.java:
```diff
@@ -889,6 +892,39 @@ public Map<String, PartitionLagState> getPartitionToLagState(
+            new SegmentErrorInfo(String.valueOf(now()), message, "")
+        );
+      }
+    } catch (TimeoutException tce) {
```
We should catch all exceptions to avoid this check breaking the ingestion (Kinesis doesn't support this method at all). You may use `fetchEarliestStreamOffset()` which already handles this

Jackie-Jiang reviewed (commented):
Please take a look at the failed tests

swaminathanmanish commented on pinot-core/src/main/java/org/apache/pinot/core/data/manager/realtime/RealtimeSegmentDataManager.java:
```diff
@@ -889,6 +892,39 @@ public Map<String, PartitionLagState> getPartitionToLagState(
+          OffsetCriteria.SMALLEST_OFFSET_CRITERIA,
+          /*maxWaitTimeMs=*/5000
+      );
+      if (streamSmallestOffset.compareTo(startOffset) > 0) {
```
Just to clarify - When the user explicitly pauses and resumes ingestion, this should not flag as a data loss right? Im assuming when we pause, segment gets committed and when ingestion is resumed from a given offset, a new segment creation kicks off from that offset and should not be seen as a gap.

vrajat replied:
If ingestion is paused and the stream fast-forwards the begin offset, then the data loss error is raised. Actually - this how I simulate this scenario in the integration test.

For my knowledge, is it possible to specify an offset when resuming the ingestion? If yes, and if the given offset < the beginOffset, the error will be raised. However that is OK ? The user expected to get messages from a specific offset and didnt.

Jackie-Jiang reviewed (commented):
Can you split the PR into 2 and focus on one thing each:
- Detect data loss and set a gauge
- Debug APIs and tests around it

Currently the PR mixes 2 things and is hard to review

vrajat commented:
Debug APIs and tests are there to test that the data loss is detected. Is it OK if one of the PRs has no test to check the changes?

Jackie-Jiang commented:
Ideally we want a unit test to directly read the value from the gauge, instead of relying on a rest API which is kind of an integration test. Currently I saw CI test failures, but not sure which part has caused it.

vrajat commented:
This PR is superseded by #12608
